wallet: Check max transaction weight in CoinSelection

aureleoules opened the pull request:
This PR is an attempt to fix #5782.

I have added 4 test scenarios, 3 of them provided here https://github.com/bitcoin/bitcoin/issues/5782#issuecomment-73819058 (slightly modified to use a segwit wallet).

Here are my benchmarks : 
## PR
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        1,466,341.00 |              681.97 |    0.6% |   11,176,762.00 |    3,358,752.00 |  3.328 |   1,897,839.00 |    0.3% |      0.02 | `CoinSelection`

## Master

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        1,526,029.00 |              655.30 |    0.5% |   11,142,188.00 |    3,499,200.00 |  3.184 |   1,994,156.00 |    0.2% |      0.02 | `CoinSelection`

achow101 commented:
It shouldn't be necessary to pass an entire CMutableTransaction to SelectCoins. CoinSelectionParams already has the information about the sizes of the rest of the transaction, modulo the change output. So a calculation can be done with that information.

Additionally, depending on the algorithm used, there may or may not be a change output, so it's size will need to be accounted for in each size calculation.

aureleoules commented:
Thank you for the feedback @achow101.
Sizes in the `CoinSelectionParams` appear to be in "bytes" and not in "weight", so I'm not sure how to use them as the multiplier of each field hasn't been taken into account, unless I'm mistaken.

achow101 commented:
The size is always vsize, so multiplyging by WITNESS_SCALE_FACTOR is sufficient to convert to weight. There will be some rounding error, but vsize always rounds up, so weight calculated in that way will always be a slight overestimate.

aureleoules commented:
I've removed the need for a dummy transaction.
Though I'm not sure how to include the change output in the size calculation yet.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -516,14 +516,32 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
+    for (const auto& result : results) {
+        int weight = coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR;
+        for (const auto& coin : result.GetInputSet()) {
+            weight += CalculateMaximumSignedInputSize(coin.txout, &wallet, &coin_control) * WITNESS_SCALE_FACTOR;
```
Can directly use `coin.input_bytes` instead of calculating the size again.

aureleoules replied:
Thanks, that's much better.

aureleoules commented:
Pinging you @Xekyo as your commented on the issue recently.

Xekyo commented:
Cool, thank you for working on this. This PR as it is would get us almost all the way there as it should ensure that any coin selection results that exceeded the weight limit get filtered out before we do our final pick out of the selection candidates. Now we'll succeed at building a transaction if there is at least one selection result that doesn't exceed the weight limit, where we might have failed previously if that solution had the best waste score (and that would be the case at a low feerate, as a higher total input weight would get preferred).

----

I was pondering whether this would fail to find an existing solution in any scenario. I came up with this:

Send ₿49.5 at a low feerate from a UXTO pool with
• 10,000 P2PKH inputs of ₿0.04
• 1,455 P2WPKH inputs of ₿0.034

Explanation:
• The UTXO pool has enough funds to pay for the transaction
• At a negligible `not_input_size`, a standard transaction can have about ~675 P2PKH inputs, but 1470 P2WPKH inputs
• With the above UTXOs, it would take 1,238 P2PKH inputs to pay for the transaction (which would exceed the limit by far), and it awould take 1,456 P2WPKH inputs to pay for the transaction (but there are too few)
⇒ Only an input set from mixed inputs that is almost exclusively composed from P2WPKH inputs (and a single digit number of P2PKH inputs) would result in a standard transaction

I don't think that any of our coin selection approaches would find such an input set, even if we added backtracking to BnB when the standard weight limit is exceeded. I figure that to succeed in this edge-case we would need an explicit mechanism to replace the inputs with the lowest value per weight once we exceed the standard weight limit. E.g. adding a heap to SRD that sorts the input set by `effectiveValue/weight`. I think this is out of scope though.


------


Approach ACK, will have to get back to review the code more carefully.

aureleoules commented:
Rebased because of silent merge conflict.

aureleoules commented:
> Send ₿49.5 at a low feerate from a UXTO pool with
> • 10,000 P2PKH inputs of ₿0.04
> • 1,455 P2WPKH inputs of ₿0.034

I tested this case with the following code and the coin selector did not find a result as you said.

```cpp
{
    // Send ₿49.5 at a low feerate from a UXTO pool with
    // 10,000 P2PKH inputs of ₿0.04
    // 1,455 P2WPKH inputs of ₿0.034

    std::unique_ptr<CWallet> wallet = std::make_unique<CWallet>(m_node.chain.get(), "", m_args, CreateMockWalletDatabase());
    wallet->LoadWallet();
    LOCK(wallet->cs_wallet);
    wallet->SetupLegacyScriptPubKeyMan();

    CoinsResult available_coins;
    for (int j = 0; j < 10000; ++j) {
        add_coin(available_coins, *wallet, CAmount(0.04 * COIN), CFeeRate(0), 144, false, 0, true);
    }

    for (int j = 0; j < 1455; ++j) {
        add_coin(available_coins, *wallet, CAmount(0.034 * COIN), CFeeRate(0), 144, false, 0, true);
    }

    const auto result = SelectCoins(*wallet, available_coins, target, cc, cs_params);
    BOOST_CHECK(result);
}
```

Xekyo commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -980,7 +980,7 @@ def test_transaction_too_large(self):
         wallet.sendmany("", outputs)
         self.generate(self.nodes[0], 10)
-        assert_raises_rpc_error(-4, "Transaction too large", recipient.fundrawtransaction, rawtx)
+        assert_raises_rpc_error(-4, "Insufficient funds", recipient.fundrawtransaction, rawtx)
```
`Insufficient funds` might be a confusing error message since the wallet would also report having sufficient funds at the same time. Perhaps we could return a more specific error, if the last attempt returns from `eligible_results` being empty e.g. `Cannot fund transaction, wallet too fragmented`.

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -518,14 +520,27 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
+        return weight <= static_cast<int>(MAX_STANDARD_TX_WEIGHT);
+    });
+
+    if (eligible_results.empty()) {
```
Perhaps we could store some “result hint” when we get into this if-block on a coin selection attempt that we could later use to inform the result message.

aureleoules replied:
Here's my attempt a3dece4e41de7bd5e118278abf09abcd2fcb4922.

murchandamus reviewed (commented):
utACK c166411dad1f4cdaa61f70b92728ecebfc3fc749

Tests look reasonable, it might be nice if we could provide a more targeted error message when the case is hit for all coin selection attempts.

---------

Yeah, I had an easier scenario at first for which I was going to suggest a fix, but then I realized that it wouldn't cause a failure because after #24584 we will always try to build input sets from a single type.
I think my above scenario is sufficiently constructed that it would be very unlikely to occur in the wild. But, since I had already tinkered it all out, I figured I might as well keep a record of a worst-case scenario.

aureleoules commented:
I have rebased this PR on top of #25734 (slightly rebased because of merge conflicts of the original PR).
Added a3dece4e41de7bd5e118278abf09abcd2fcb4922 to show a more specific error message when the transaction weight is exceeded during the coin selection.

furszy commented:
For the specific error messages, I don't think that passing an enum reference to the function is the best solution. We have the `util::Result` class to cover this exact purpose (more info in #25218 and #25721)

Plus, some notes about this in case you haven't seen them:

* I decoupled the preset inputs fetching block of code from `SelectCoins` and added `util::Result` class connection with proper error descriptions in #25685.
(the error descriptions there are not due "Insufficient Funds", that was wrong).

* And then have #25806, which has 3b3054f9 which basically adds `util::Result` error support to the entire `SelectCoins` function (which is what you actually need here).

So, would suggest to either cherry-pick 3b3054f9 here, so you can return the error string (returning an empty string if the error is "Insufficient Funds" so there aren't changes out of the scope of this PR and you can still return the specific error) or, if you don't want it, maybe leave the proper error description point for a follow-up PR and only focus on the functionality here.

aureleoules commented:
Thank you @furszy for you comments!
I'll probably wait until #25806 and #25734 are merged before continuing working on this error message commit and will drop a3dece4e41de7bd5e118278abf09abcd2fcb4922 in the meantime.
I rebased it again since #25734 just got rebased.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -531,14 +533,27 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
 
+    std::vector<SelectionResult> eligible_results;
+    std::copy_if(results.begin(), results.end(), std::back_inserter(eligible_results), [coin_selection_params](const SelectionResult& result) {
+        const int weight = std::accumulate(result.GetInputSet().begin(), result.GetInputSet().end(), static_cast<int>(coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR), [](int acc, const auto& coin) {
```
`tx_noinputs_size` is 0 when we are subtracting the fee from the outputs, so using it here may not always result in the correct calculation.

aureleoules replied:
Do you know how I can take this into account? I did hear you say "fuck" on your stream VOD while reviewing this so I guess that's not trivial?

achow101 replied:
Essentially you need to find everywhere that `tx_noinputs_size` is being set/updated and make sure that the places that use `tx_noinputs_size` for calculations treat it as if it were 0 when SFFO. I don't think this should be too difficult as it is only used within `CreateTransactionInternal` and only for calculating `not_input_fees`.

Xekyo replied:
I am not surprised that `not_input_fees` would be zero in case of SFFO, but quite surprised that `tx_noinputs_size` would be 0, though. Just given the variable name, I would have expected it to simply be the size of the recipient outputs plus header data, regardless of who pays the fees. Would it perhaps be doable to fix the behavior in that direction, to reduce future developer surprise?

aureleoules replied:
I applied your advice @achow101 here: https://github.com/bitcoin/bitcoin/compare/32a233773b4536aadecd00a52c897614dc9981b8..9421f362d5e0c6d4c7fe047a23ef7e0676f380b9.
Is this right?

Also, maybe we could make `tx_noinputs_size` private and have a getter that returns 0 in case of SFFO?

achow101 replied:
No, that's not what I meant.

What you've done is no different from how it was before. My point was that you need to be using the non-inputs size always, not set it to 0 for SFFO. This size is typically tracked by `tx_noinputs_size`, but it gets set to 0 when SFFO. So you need to find everywhere else (generally in `CreateTransactionInternal`) that `tx_noinputs_size` is used, change it so that it is always calculated, and change the places that expected it to be 0 for SFFO to still use 0 (e.g. with a ternary).

aureleoules replied:
Ah okay my bad. I pushed it, should be good?

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -917,5 +917,176 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+        /*effective_feerate=*/CFeeRate(0),
+        /*long_term_feerate=*/CFeeRate(0),
+        /*discard_feerate=*/CFeeRate(0),
+        /*tx_noinputs_size=*/static_cast<size_t>(10) + GetSerializeSize(txout, PROTOCOL_VERSION), // static header size + output size
```
Since sizes are already being hardcoded above for `change_output_size` and `change_spend_size`, I think it's fine to hard code this as well instead of having to create an entire wallet just to fetch a single destination in order to get this size, where accuracy also doesn't really matter.

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -917,5 +917,176 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+    FastRandomContext rand;
+    CoinSelectionParams cs_params{
+        rand,
+        /*change_output_size=*/34,
+        /*change_spend_size=*/148,
```
While these values don't particularly matter, they should probably match the p2wpkh sizes rather than the p2pkh.

```suggestion
        /*change_output_size=*/31,
        /*change_spend_size=*/68,
```

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -917,5 +917,176 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+        std::unique_ptr<CWallet> wallet = std::make_unique<CWallet>(m_node.chain.get(), "", m_args, CreateMockWalletDatabase());
+        wallet->LoadWallet();
+        LOCK(wallet->cs_wallet);
+        wallet->SetupLegacyScriptPubKeyMan();
```
Let's avoid making LegacySPKMs in tests which are not specifically testing that behavior.

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -917,5 +917,176 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+    txDummy.vout.push_back(txout);
+
+    {
+        // Scenario 1:
```
ISTM these tests could be deduplicated intoa helper function with each scenario just being a set of parameters as each test case does basically the same thing but with slightly different params.

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -917,5 +917,176 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+    }
+
+    {
+        std::unique_ptr<CWallet> wallet = std::make_unique<CWallet>(m_node.chain.get(), "", m_args, CreateMockWalletDatabase());
+        wallet->LoadWallet();
+        LOCK(wallet->cs_wallet);
+        wallet->SetupLegacyScriptPubKeyMan();
+
+        CoinsResult available_coins;
+        for (int j = 0; j < 3000; ++j) {
+            add_coin(available_coins, *wallet, CAmount(0.017 * COIN), CFeeRate(0), 144, false, 0, true);
+        }
+
+        // Perform selection
+        const auto result = SelectCoins(*wallet, available_coins, target, cc, cs_params);
+        BOOST_CHECK(!result);
```
A comment for this test case would be nice.

aureleoules commented:
I refactored the tests and changed the values a bit because I switched from a legacy wallet to a descriptor wallet.
Also documented the last test.

aureleoules committed:
wallet: Check max tx weight in coin selector

Co-authored-by: Andrew Chow <github@achow101.com>

aureleoules committed:
test: Check max transaction weight in CoinSelection

Co-authored-by: Andrew Chow <github@achow101.com>

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -511,14 +513,28 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
 
+    std::vector<SelectionResult> eligible_results;
+    std::copy_if(results.begin(), results.end(), std::back_inserter(eligible_results), [coin_selection_params](const SelectionResult& result) {
+        const auto initWeight{coin_selection_params.m_subtract_fee_outputs ? 0 : coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR};
```
In dbeddc1f7f43c6df293133cd2b12a7187c0e321e "wallet: Check max tx weight in coin selector"

We still need to calculate the weight correctly even when subtracting the fee from outputs, so this should not be checking that.

```suggestion
        const auto initWeight{coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR};
```

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -922,6 +922,169 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+        const bool has_0_0625_coin = std::any_of(result->GetInputSet().begin(), result->GetInputSet().end(), [](const auto& coin) {
+            return coin.GetEffectiveValue() == 0.0625 * COIN;
+        });
+        const bool has_not_50_coin = std::any_of(result->GetInputSet().begin(), result->GetInputSet().end(), [](const auto& coin) {
+            return coin.GetEffectiveValue() != 50 * COIN;
+        });
```
In fc5219f51813b6e9c128d6bc0cc70c63a5edce48 "test: Check max transaction weight in CoinSelection"

It's not clear to my why we shouldn't expect the 50 Coin to be picked? ISTM Random selection should eventually find the 50 Coin single input solution and go with that.

aureleoules replied:
Right, I removed the check. The two other coins should always be present though.

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -922,6 +922,169 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+        const auto result = select_coins(
+            target, cs_params, cc, [&](CWallet& wallet) {
+                CoinsResult available_coins;
+                for (int j = 0; j < 1500; ++j) {
```
In fc5219f51813b6e9c128d6bc0cc70c63a5edce48 "test: Check max transaction weight in CoinSelection"

The comment says 1515, but this is only 1500? It seems like 1500 works anyways.

aureleoules commented:
Thanks @achow101 I addressed your comments and added you as co-author.

aureleoules commented:
Rebased and solved silent merge conflicts.

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -930,6 +931,157 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+    }
+
+    {
+        // Scenario 3 (available as RPC test):
+
+        // The actor starts with 1x 50.0 BTC, 400x 0.0625 BTC and 2000x 0.025 BTC (125.0 BTC total) unspent outputs
+        // Then tries to spend 49.5 BTC
+        // A combination of coins should be selected, such that the created transaction is not too large
+
+        // Perform selection
+        const auto result = select_coins(
+            target, cs_params, cc, [&](CWallet& wallet) {
+                CoinsResult available_coins;
+                for (int j = 0; j < 400; ++j) {
+                    add_coin(available_coins, wallet, CAmount(0.0625 * COIN), CFeeRate(0), 144, false, 0, true);
+                }
+                for (int j = 0; j < 2000; ++j) {
+                    add_coin(available_coins, wallet, CAmount(0.025 * COIN), CFeeRate(0), 144, false, 0, true);
+                }
+                add_coin(available_coins, wallet, CAmount(50 * COIN), CFeeRate(0), 144, false, 0, true);
+                return available_coins;
+            },
+            chain, m_args);
+
+        BOOST_CHECK(has_coin(result->GetInputSet(), CAmount(0.025 * COIN)));
+        BOOST_CHECK(has_coin(result->GetInputSet(), CAmount(0.0625 * COIN)));
+    }
```
In 54ecb044c0ebf40ec06bd192725d6f49dcff90ff "test: Check max transaction weight in CoinSelection"

It's not clear to me how this scenario is materially different from scenario 2. It just adds the 50 Coin UTXO and I'm not sure what that actually adds to the test?

Additionally, this will fail 1 in 2401 runs as the 50 Coin could be chosen as a single input solution. This would then fail the checks.

aureleoules replied:
Thanks, I removed it.

aureleoules commented:
Removed scenario 3 and removed unnecessary check in scenario 2 (https://github.com/bitcoin/bitcoin/compare/54ecb044c0ebf40ec06bd192725d6f49dcff90ff..ffd744f9a0391beda6bb98de1d4e851fc6ca7ca4)

aureleoules commented:
Rebased to fix CI `p2p_sendtxrcncl.py` failure.

achow101 commented:
ACK b763b072cec235d397fae4dafc6a69f5902eb3f6

furszy commented on src/wallet/spend.cpp:
```diff
@@ -557,14 +559,28 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
 
+    std::vector<SelectionResult> eligible_results;
+    std::copy_if(results.begin(), results.end(), std::back_inserter(eligible_results), [coin_selection_params](const SelectionResult& result) {
+        const auto initWeight{coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR};
+        const int weight = std::accumulate(result.GetInputSet().begin(), result.GetInputSet().end(), static_cast<int>(initWeight), [](int acc, const auto& coin) {
+            return acc + coin.input_bytes * WITNESS_SCALE_FACTOR;
+        });
```
Instead of doing this entire for-loop accumulation, what about adding a field `m_weight_sum` to `OutputGroup`?

Initializing it to `coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR` in the constructor and increasing it at every output insertion (inside `OutputGroup::Insert`).

So, when `SelectionResult::AddInput` gets called, you already have the total weight for each output group. And can just sum it up inside another cache field there.

aureleoules commented:
Thanks @furszy, I've implemented your suggestion: https://github.com/bitcoin/bitcoin/compare/b763b072cec235d397fae4dafc6a69f5902eb3f6..119fea975463b05d9a2dddbb36db17b2c91ef880.

It appears to be slightly faster:

### Before (dc9bce1866f1cce851a216d2c7bfd7a890c27006)
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        1,614,196.00 |              619.50 |    0.6% |   11,151,218.00 |    3,710,976.00 |  3.005 |   1,995,456.00 |    0.3% |      0.02 | `CoinSelection`

### Current (98efb22650c803c3bd4c094d58ef7ba16b77fee2)
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        1,466,341.00 |              681.97 |    0.6% |   11,176,762.00 |    3,358,752.00 |  3.328 |   1,897,839.00 |    0.3% |      0.02 | `CoinSelection`

I've updated the PR description's benchmarks.

furszy commented:
> I also realized that in the previous commit the weight could go in the negatives because coin.input_bytes is equal to -1 for external inputs and I simply summed this number, I fixed it.

Hmm, that shouldn't be happening. We do have a check inside the preset inputs fetching function that throws an error if `input_bytes == -1`. Maybe what your wallet is trying to spend is a not solvable coin owned by the wallet.

aureleoules commented:
Sorry @furszy I meant that the weight could be negative in other tests, which probably used external coins. My test was not affected.

achow101 commented:
ACK 3474b5f7684cacd0c098a11c95079e29f5e905b4

aureleoules commented:
Thanks @achow101. @furszy suggested to me in private that the WITNESS_SCALE_FACTOR may not be correctly applied tough. I drafted a quick diff here: 
https://gist.github.com/aureleoules/8a7cfcd93a66b70eecc74e7a6692d300

Would you agree?

achow101 commented:
No, it is correct as it is now. Converting vbytes to weight is always just multiple by WITNESS_SCALE_FACTOR. It doesn't matter what the script type is.

aureleoules commented:
Okay great thank you!

furszy commented on src/wallet/coinselection.cpp:
```diff
@@ -436,18 +441,25 @@ void SelectionResult::Clear()
     m_use_effective = !subtract_fee_outputs;
+
+    m_weight += std::accumulate(inputs.cbegin(), inputs.cend(), 0, [](int sum, const auto& coin) {
+        return sum + std::min(coin.input_bytes, 0) * WITNESS_SCALE_FACTOR;
```
Should be `std::max(coin.input_bytes, 0)`.

Otherwise, this entire sum will be 0 most of the time (it's taking the smaller value between `input_bytes` and 0).

aureleoules replied:
Right thanks, I confused myself. Pushed.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -557,14 +559,24 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
     }
 
-    if (results.size() == 0) {
+    if (results.empty()) {
         // No solution found
         return std::nullopt;
     }
```
As `std::copy_if` is a no-op if the vector is empty, could remove these lines.

aureleoules replied:
Leaving the `results.empty())` check would also avoid creating the `eligible_results` no?

furszy replied:
yes, would only avoid the empty vector initialization.

furszy commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -930,6 +931,124 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+            chain, m_args);
+
+        BOOST_CHECK(result);
+        BOOST_CHECK(has_coin(result->GetInputSet(), CAmount(50 * COIN)));
```
As this is expecting to find one single output, might be better to do:

```c++
const auto& inputs = result->GetInputSet();
BOOST_CHECK_EQUAL(inputs.size(), 1);
BOOST_CHECK_EQUAL(inputs.begin()->GetEffectiveValue(), 50 * COIN);
```

furszy replied:
hmm, privacy wise, sounds odd to add hundreds of UTXO from different external sources when the target amount is being covered only by one of them.. the rest are not needed, they are just leaking information about the wallet and making the tx bigger.

Will give it a look. It's not something that should block this PR, just a parallel investigation.

furszy reviewed (approved):
code review ACK 30b71d1f, left few small comments.

achow101 commented:
ACK 30b71d1f0ea7e8b69b069713b66e45d472373832

aureleoules commented:
Rebased

achow101 commented:
reACK c7c7ee9d0b2d7b303b9300f941e37e09e7d8d8b6

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/25729/commits/c7c7ee9d0b2d7b303b9300f941e37e09e7d8d8b6

S3RK commented on src/wallet/coinselection.cpp:
```diff
@@ -354,6 +355,10 @@ void OutputGroup::Insert(const COutput& output, size_t ancestors, size_t descend
     m_descendants = std::max(m_descendants, descendants);
+
+    if (output.input_bytes > 0) {
+        m_weight += output.input_bytes * WITNESS_SCALE_FACTOR;
```
While this is fine for the purposes of this PR, the weight here is not always precise due to rounding error of `input_bytes`.

Ideally, we would store the actual weight in `COutput` and just take it from there. I realise, this might be out of scope for this PR. But maybe we can add a warning in code for future devs and create an issue in this repo?

w0xlt replied:
> the weight here is not always precise due to rounding error of input_bytes

Could elaborate, since both `COutput::input_bytes` and `WITNESS_SCALE_FACTOR` are integer ?

achow101 replied:
1 virtual byte = 4 weight units. When computing vbytes from weight, if the weight is not divisible by 4, the vbytes is rounded up. For example, 6 WU is really 1.5 vbytes, but is rounded to 2 vbytes. Since `input_bytes` stores this rounded number, computing weight from it results in 8 WU rather than the correct value of 6.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -888,7 +897,7 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
 
     // Include the fees for things that aren't inputs, excluding the change output
-    const CAmount not_input_fees = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.tx_noinputs_size);
+    const CAmount not_input_fees = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.m_subtract_fee_outputs ? 0 : coin_selection_params.tx_noinputs_size);
```
could:
```suggestion
    const CAmount not_input_fees = coin_selection_params.m_subtract_fee_outputs ? 0 : coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.tx_noinputs_size);
```

achow101 replied:
I would actually rather that `not_input_fees` is always calculated, and the SFFO handling is done for just the target. We really need to try to reduce the number of places that we have differences between normal and SFFO coin selection. Since `not_input_fees` is used elsewhere, having its value be dependent on SFFO results in a cascading effect that I think we should be avoiding.

But that can be done in a followup.

furszy reviewed (approved):
diff ACK c7c7ee9d
